### PR TITLE
Add support for fetching and saving external model definitions

### LINKED ISSUE #974: Export models in standardized "archival" format
RV model bundles contain RV configuration which is used to recreate the model architecture at runtime. This works if we are using the model within RV, but makes it harder to use in other settings. We would like to also export the model in a more "archival" format so it can be instantiated outside of RV. @AdeelH researched this and found that TorchHub might be a good candidate for a way to do this https://pytorch.org/docs/stable/hub.html

Before starting work on this, we should have a discussion to see if this is the best approach and how we would add support for it in RV.

### LINKED ISSUE #975: Add Support for External Architectures
Adding easy support for externally-defined architectures seems as though it might be an interesting feature to have.  In particular, I am interested in being able to

1. point raster vision to some architecture definition code (in some understood, canonical format) and train a model that uses that architecture (see #927 for a proto-example), and
2. generate a model bundle that contains the architecture definition code (or possibly a reference to it) within itself in addition to the model weights (relates to #974, though that issue seems to contemplate doing this for existing raster vision architectures [which I am also in favor of])

@AdeelH discovered [torch hub](https://pytorch.org/docs/stable/hub.html) which looks as if it provides a clean way to accomplish the loading and referencing of externally-defined architectures.  Some questions that came up while reading through the documentation: is it possible to reference a particular commit on GitHub? (Branches and tags are both mentioned, but untagged commits are not [even if this is not supported it is not a deal-breaker for me].)  Also, there may be instances when it is convenient to save the actual architecture code in the bundle rather than a reference to it: that does not seem to be supported.

Tagging @AdeelH @lewfish

### PR DESCRIPTION
**Pull request by AdeelH**
## Overview

Adds support for fetching and saving external model definitions (#975). Currently, only supported for chip classification and semantic segmentation.

- External model definitions can be specified by supplying an `ExternalModuleConfig` to `ModelConfig.external_def`.
- Example usage:
	- classification (`spacenet_rio.py`)
	```python3
	model = ClassificationModelConfig(
		external_def=ExternalModuleConfig(
			github_repo='lukemelas/EfficientNet-PyTorch',
			# uri='s3://raster-vision-ahassan/models/EfficientNet-PyTorch.zip',
			name='efficient_net',
			force_reload=False,
			entrypoint='efficientnet_b0',
			entrypoint_kwargs={
				'num_classes': len(class_config.names),
				'pretrained': 'imagenet'
	}))
	```
	- semantic segmentation
	```python3
    model = SemanticSegmentationModelConfig(
		external_def=ExternalModuleConfig(
			github_repo='pytorch/vision',
			name='fcn',
			force_reload=False,
			entrypoint='fcn_resnet50',
			entrypoint_kwargs={
				'num_classes': len(class_config.names),
				'pretrained': False
	}))
	```
- ~New example file: `examples/chip_classification/spacenet_rio_external_model.py`~ Example usage included in `spacenet_rio.py`, and can be enabled through the `external_model` argument.
- The `learner` uses this config to fetch the code via torch.hub.
- The definition is stored under `<output_uri>/train/modules`.
- During the bundling step, `<output_uri>/train/modules` is copied to the bundle directory and thereby included in the bundle zip file.
- When loading from a bundle, the definition is recovered and its path provided to `learner.__init__` via a new `model_def_path` param.
- The `torch.hub` wrappers can be found at `.../pytorch_learner/utils/torch_hub.py`
- `.../pytorch_learner/utils.py` has been moved to `.../pytorch_learner/utils/utils.py`

### Checklist

- [x] Updated `docs/changelog.rst`
- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [x] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

## Testing Instructions

- How to test this PR
    - Use examples above. Try with different options.

Closes #975

Makes progress towards #974

### CODE REVIEW

**Comment by AdeelH** [comment]
This does not solve #974 directly, but it should now be possible to decouple internal model definition code from the learners and move it all to a separate place, like a new `pytorch_modules` plugin, or a new repo, and use this mechanism to load them.

**Review by lewfish — CHANGES_REQUESTED** [review]
Looks good overall.

File: rastervision_pytorch_backend/rastervision/pytorch_backend/examples/chip_classification/spacenet_rio_external_model.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    chip_sz = 200
+    train_scenes = [make_scene(info) for info in train_scene_info]
+    val_scenes = [make_scene(info) for info in val_scene_info]
+    dataset = DatasetConfig(
+        class_config=class_config,
+        train_scenes=train_scenes,
+        validation_scenes=val_scenes)
+
+    model = ClassificationModelConfig(
```
[the remark this answers is not in the document]
Comment (AdeelH) [reply rastervision_pytorch_backend/rastervision/pytorch_backend/examples/chip_classification/spacenet_rio_external_model.py]: Done. Added an `external_model` flag to `get_config()`.

```python3
    if external_model:
        model = ClassificationModelConfig(
            external_def=ExternalModuleConfig(
                github_repo='lukemelas/EfficientNet-PyTorch',
                ...
                }))
    else:
        model = ClassificationModelConfig(backbone=Backbone.resnet50)
```

File: rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_learner_backend_config.py [hunk 2]
```diff
@@ -50,3 +50,16 @@ def validate_solver_config(cls, v):
                     'class_loss_weights is currently only supported for '
                     'Semantic Segmentation and Chip Classification.')
         return v
+
+    @validator('model')
+    def validate_model_config(cls, v):
+        if v.external_def is not None:
```
[the remark this answers is not in the document]
Comment (AdeelH) [reply rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_learner_backend_config.py]: I have changed it so that the object detection backend checks if `ModelConfig.external_def` has been set and raises an error if it has.

**Review by lewfish — APPROVED** [review]

**Comment by jamesmcclain** [comment]
:tada:
